backend/api/routers/coupon.py

```diff
@@ -104,13 +107,22 @@ def delete_coupon(
     if coupon.time_used is not None:
         raise HTTPException(status_code=409, detail="Coupon already used")
 
+    role = client_info.get("sub")
+    username = client_info["name"]
+    is_staff = role in ["manager", "exchange"]
+    is_client = role is None or role == "client"
+    if (
+        is_client and recently_issued_coupons.get(serial_num) != username
+    ) or not is_staff:
+        raise HTTPException(status_code=409, detail="Authorization error")
+
```

Comment:
Potential security vulnerability
Since ownership is checked after checking if coupon exists, it is possible to find unused coupons by bruteforce

Reply:
it's ok, we can just ratelimit for future

Reply:
There is no nice way to fix this without breaking requirements, but can be prevented with rate limit